clang/lib/CodeGen/CodeGenModule.cpp

```diff
@@ -6235,6 +6235,19 @@ void CodeGenModule::EmitGlobalVarDefinition(const VarDecl *D,
     getCUDARuntime().handleVarRegistration(D, *GV);
   }
 
+  if (LangOpts.isSYCL()) {
+    const RecordDecl *RD = D->getType()->getAsRecordDecl();
+    if (RD && RD->hasAttr<SYCLDeviceGlobalAttr>()) {
+      // SYCL device globals are initialized externally
+      GV->setExternallyInitialized(true);
+      // Since static device global symbols need to cross host device boundary,
+      // don't allow internal linkage as it might tamper with the device global
+      // symbol
+      if (Linkage == llvm::GlobalValue::InternalLinkage)
+        Linkage = llvm::GlobalValue::ExternalLinkage;
```

Comment:
External linkage here can mean this symbol can interpose one in another translation unit. What happens when these two translation units are linked by someone expecting standard C++ linkage rules to apply:

a.cpp
```cpp
static device_global<int[4], somepTestProperties> foo;
...
```

```
b.cpp
double *foo;
...
```
I don't have an answer to the above question, but I don't think marking the symbol as `ExternalLinkage` is going to be right on its own.

Reply:
This is the same linkage as is used in the CUDA language in LLVM. I think problems of the same symbols across different TUs should be OK as this `EmitGlobalVar` call will only get to this point for the SYCL device compilation pass, not for the host pass. The host compilation honours normal `static` behaviour. I'll add some testing to make sure this is happening.

Reply:
I have restricted changes so the linkage is only changed for NVPTX and AMDGCN backends, as I don't know if it's a good idea to do so for other backends. This makes the linkage equivalent for CUDA device globals/constants and SYCL device globals.